Blob id v2

xiahome opened the pull request:
This patch introduces BlobId v2 that embeds accountId, containerId, and datacenterId. The first two will be an in-blobId identifier for the ownership of the blob, and the third will be used to get the datacenter where the blob was originally posted.

To include the three additional pieces of information, the blobId version needs to be increased from 1 to 2. To better support multiple versions, a builder class is introduced to build blobIds in different versions.

Test done:
./gradlew clean && ./gradlew build

Test coverage:
Class Class, %  Method, % Line, %
BlobId  100% (1/ 1) 88.2% (15/ 17)  80.4% (78/ 97)
BlobIdBuilder 100% (1/ 1) 100% (7/ 7) 100% (16/ 16)

fixes linkedin/ambry#636

pnarayanan commented:
Minor: Could you update the commit message with the phrase `fixes linkedin/ambry#636` ? That ensures that the associated issue gets closed when this PR is merged.

pnarayanan commented:
Also - based on our discussions with @asinghai, let us also add an additional byte to the id for future use?

nsivabalan commented:
@xiahome : is this ready to review or you wanna add the byte field ?

xiahome commented:
@nsivabalan Per offline discussion, will add a byte as the flag.

xiahome commented:
@pnarayanan @nsivabalan @asinghai Have made the following changes:
1. add a byte for flag;
2. use a byte to carry datacenter id.

nsivabalan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -66,48 +156,89 @@ public BlobId(String id, ClusterMap clusterMap) throws IOException {
 
+  /**
+   * Gets the id of the {@link Account} who created this blob. If this information was not available when the
+   * blob id was formed, it will return -1.
```
-1 only in case of older versioned blobId. May be good to call it out explicitly.

xiahome replied:
The point is that `-1` will be used across `Ambry` to indicate the value was not available or not assigned at the time when the `blobId` was formed. To the caller, this is just a numerical flag.

nsivabalan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -124,6 +255,20 @@ public String getLongForm() {
+
+      default:
+        throw new IllegalArgumentException("blob id version=" + version + " not supported");
+    }
```
a question about compareTo() methods that is untouched. Shouldn't we compare getID() and not individual partitionId and uuid? StoreKey comparison is for the purpose of index entries right.

xiahome replied:
`getID()` is a bit expensive, not sure if this was the original reason not use it. Following the same way I could add `flag`, `datacenterId`, `accountId`, and `containerId`.

nsivabalan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobIdBuilder.java:
```diff
@@ -0,0 +1,111 @@
+   * @param partitionId The {@link PartitionId} to be set to the blob id.
+   */
+  public BlobIdBuilder(Byte flag, Byte datacenterId, Short accountId, Short containerId, PartitionId partitionId) {
+    this.flag = flag;
```
do you think we need this constructor too?  can't the callers use set methods ?

pnarayanan replied:
+1.

xiahome replied:
It just makes less code, and most of the time the five arguments can be available so no need to set individually.
```
    BlobId blobId = new BlobIdBuilder(referenceFlag, referenceDatacenterId, referenceAccountId, referenceContainerId,
        referencePartitionId).build();
```
vs.
```
    blobId = new BlobIdBuilder(referencePartitionId).setFlag(referenceFlag)
        .setDatacenterId(referenceDatacenterId)
        .setAccountId(referenceAccountId)
        .setContainerId(referenceContainerId)
        .build();
```

nsivabalan replied:
we should have just one constructor with all mandatory fields. For optional fields, callers should resort to using setters.

xiahome replied:
Ok, I just removed this constructor.

nsivabalan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobIdFactory.java:
```diff
@@ -33,6 +33,6 @@ public BlobIdFactory(ClusterMap clusterMap) {
-    return new BlobId(value, clusterMap);
+    return BlobId.fromDataInputStream(value, clusterMap);
   }
 }
```
new line

xiahome replied:
Fixed.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -162,18 +307,52 @@ public boolean equals(Object o) {
+   * {@link #BLOB_ID_V1}.
+   */
+  private void populateMissingFieldsAndDetermineVersion() {
+    version = flag == null && accountId == null && containerId == null && datacenterId == null ? BlobId.BLOB_ID_V1
```
@nsivabalan @pnarayanan I am considering if we should regard default values the same as not set.

pnarayanan replied:
So here is the thing. This change will have to go in the same way as other version changes. All readers should know how to deserialize a V2 BlobId. You will also have to ensure that once you switch to V2 a node does not go back to V1. We cannot have V1 and V2 blobs go in at the same time as the store will keep creating new index segments due to the size change, which is not good.

Please keep that in mind when making this change. It might mean that you keep a "Current_BlobId_Version" which is set to V1 and at some point you switch to V2 and at that point even if these fields are null you will populate it with -1. 

I am wondering how we should deal with this during deployment (in the interim period when one frontend writes in V1 and another in V2 and they both put to the same server). 

Let us think this through and flesh this out properly.

nsivabalan replied:
This should be same as BlobProperties. Server always deals with storeKeys. So, in first version we will ensure all servers and frontends will be able to read BlobIdV2. new BlobId() will be called only by router. So, when we start deploying 2nd version, FE_V1 will send out PUTs in BlobIdV1 and FE_V2 will send out PUTs in BlobIdV2. Since servers will be able to read both blobId versions, they should be able to handle both type of requests (from both FE_V1 and FE_V2)

nsivabalan replied:
But the tricky part here is the index roll over. during this deployment, during replication some servers might send blobIds in V1 and some in blobIds in V2. If one server receives interleaved blobId  versions from same or multiple server, this will cause roll over of indexes for every size diff between successive storeKey(aka blobIds).

xiahome replied:
Thanks for comments from both of you, and it seems a common point is that we should be careful when enabling the write path. Let me clarify two things:
1. after this pr, both blob id v1 and v2 can be understood;
2. this pr does not enable write blob id v2. Write is actually enabled by router depending on how router uses the `BlobIdBuilder`. In the original version of this pr, when router supplies one of the four required fields by v2 (`flag`, `datacenterId`, `accountId`, and `containerId`), it will enable v2.

Based on the second point, router has the control to write v2. However, as discussed with Priyesh, we want to make this safe and unambiguous, so now a `CURRENT_VERSION` is introduced into `BlobId`. When this is set to v1, it will construct blobId v1 regardless how we set the four fields required by v2. Once we decide to write v2 at router side, we also need to change `CURRENT_VERSION` to v2. This will gives us more confidence.

Because of this change, I have rewritten a significant part of the unite tests, because the logic of deciding version is not necessary to test.

pnarayanan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,28 +25,116 @@
 public class BlobId extends StoreKey {
-  private Short version = 1;
-  private static short Version_Size_In_Bytes = 2;
+  static final short BLOB_ID_V1 = 1;
```
these can be private?

xiahome replied:
Tests heavily used it. Do you think it makes sense to make it `protected` and extend the `BlobId` class?

pnarayanan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -153,7 +298,7 @@ public boolean equals(Object o) {
       return true;
     }
-    if (o == null || getClass() != o.getClass()) {
+    if (!(o instanceof BlobId)) {
```
why make this change (and the original way is safer, I think).

xiahome replied:
This is auto-generated by IntelliJ. I can revert it back if needed.

pnarayanan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -162,18 +307,52 @@ public boolean equals(Object o) {
     if (!version.equals(blobId.version)) {
```
why not simply write `return version.equals(blobId.version) && partitionId.equals(blobId.partitionId) && .... && uuid.equals(blobId.uuid)`?

xiahome replied:
This is lazily done by using the IntelliJ auto generation.

xiahome commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/Datacenter.java:
```diff
@@ -31,7 +31,9 @@
-class Datacenter {
+public class Datacenter {
+  // datacenterId = LEGACY_DATACENTER_ID indicate datacenterId is not available at the time when this blobId is formed.
+  public static final byte LEGACY_DATACENTER_ID = -1;
```
Moved.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -14,6 +14,7 @@
 package com.github.ambry.commons;
 
 import com.github.ambry.clustermap.ClusterMap;
+import com.github.ambry.clustermap.Datacenter;
```
This dependency is no longer there.

xiahome commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/CompositeClusterManager.java:
```diff
@@ -23,6 +23,8 @@
 import java.util.List;
 import java.util.Set;
 
+import static com.github.ambry.clustermap.Datacenter.*;
```
Now there is no dependency on `Datacenter` after the `LEGACY_DATACENTER_ID` is moved to `ClusterManagerUtils`.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/Account.java:
```diff
@@ -31,6 +31,8 @@
  * is made in JSON, which is generic to contain additional information of the metadata.
  */
 public class Account {
+  // accountId = LEGACY_ACCOUNT_ID indicate accountId is not available at the time when this blobId is formed.
```
Fixed and the same changes made to `accountId` and `containerId`.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,28 +25,116 @@
+ *   BlobId uniquely identifies a stored blob. A blobId is a reference that is returned back to a caller when
+ *   posting a blob, and later will be required to fetch the blob.
+ * </p>
+ * <p>
```
That makes sense, have added.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,28 +25,116 @@
+ * </p>
+ * <pre>
+ * +---------------------------------------------+
+ * | version | partitionId | uuidSize | uuid     |
```
Fixed.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,28 +25,116 @@
+ * </p>
+ * <pre>
+ * +---------+-------+--------------+-----------+-------------+-------------+----------+----------+
+ * | version | flag  | datacenterId | accountId | containerId | partitionId | uuidSize | uuid     |
```
Fixed.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,28 +25,116 @@
+  static final short BLOB_ID_V2 = 2;
+  static final byte DEFAULT_FLAG = 0;
+  private static final short VERSION_SIZE_IN_BYTES = 2;
+  private static final int UUID_SIZE_IN_BYTES = 4;
```
I do not know the reason, should we change it to short since blobIdv2?

xiahome replied:
This is a good suggestion. Have made changes accordingly.

xiahome replied:
Had an offline discussion, will keep this as short to not introduce confusion between v1 and v2 on the same field.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,28 +25,116 @@
+   * @param containerId The id of the {@link Container} to be embedded into the blob.
+   * @param partitionId The partition where this blob is to be stored.
+   */
+  BlobId(Byte flag, Byte datacenterId, Short accountId, Short containerId, PartitionId partitionId) {
```
It is called by the `Builder`, and now we have the style that `Builder` and the actual component are in different classes.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -66,48 +156,89 @@ public BlobId(String id, ClusterMap clusterMap) throws IOException {
    */
-  public BlobId(DataInputStream stream, ClusterMap clusterMap) throws IOException {
-    this(stream, clusterMap, false);
+  public static BlobId fromDataInputStream(DataInputStream stream, ClusterMap clusterMap) throws IOException {
```
On the other day I had a discussion with @nsivabalan . The way we are going to adopt is: 1) using `builder`s to build objects that take arguments; and 2) using static methods to deserialize a stream/json/file etc. into an object. If these are the only two ways to get an object, then the constructors should be made at least package-private.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -66,48 +156,89 @@ public BlobId(String id, ClusterMap clusterMap) throws IOException {
+   * will return 0.
+   * @return The flag of the blobId.
+   */
+  public short getFlag() {
```
Have added in the javadoc of `BlobId` class, above the blobid format.

nsivabalan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,28 +25,116 @@
     }
-    this.partitionId = partitionId;
-    this.uuid = UUID.randomUUID().toString();
+    uuid = Utils.readIntString(stream);
```
I see that you have changed the UUID_SIZE to short. Then this has to be short followed by a string right.

nsivabalan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -66,48 +156,89 @@ public BlobId(String id, ClusterMap clusterMap) throws IOException {
+        throw new IllegalArgumentException("blob id version=" + version + " not supported");
+    }
     idBuf.put(partitionId.getBytes());
     idBuf.putInt(uuid.getBytes().length);
```
You are putting int as UUID_SIZE but the doc says otherwise

xiahome replied:
Have changed toe doc to short. After discussion with Priyesh, it was decided to keep using int not to introduce confusion on the same field used in two versions.

xiahome commented:
Have updated the pr with a new commit addressing the comments.

nsivabalan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,90 +27,251 @@
+ *   returned back to a caller when posting a blob, and later will be required to fetch the blob.
+ * </p>
+ * <p>
+ *   There are two versions of format for blob de/serialization. Version 1, which includes {@code partitionId}
```
typo. "blobId de/ser" instead of "blob de/ser"

xiahome replied:
Fixed.

nsivabalan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,90 +27,251 @@
+  static final short BLOB_ID_V2 = 2;
+  static final byte DEFAULT_FLAG = 0;
+  private static final short CURRENT_VERSION = BLOB_ID_V1;
+  private static final short VERSION_FIELD_LENGTH_IN_BYTES = 2;
```
can we start using Short.BYTES, Integer.BYTES and so on to fetch primitive sizes

xiahome replied:
Changed as suggested.

nsivabalan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,90 +27,251 @@
+  // the version to indicate the serialized format.
+  private Short version;
+  private Byte flag;
+  private Byte datacenterId;
```
why are these not final ?

xiahome replied:
They cannot be declared as final because we have the logic to populate missing fields.

nsivabalan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -140,9 +318,21 @@ public int compareTo(StoreKey o) {
-        result = uuid.compareTo(other.uuid);
+        result = datacenterId.compareTo(other.datacenterId);
+        if (result == 0) {
+          result = accountId.compareTo(other.accountId);
```
you could also do it as multiple if loops.
```
    if (result == 0) {
      result = flag.compareTo(other.flag);
    }
    if (result == 0) {
      result = datacenterId.compareTo(other.datacenterId);
    }
    if (result == 0) {
      result = accountId.compareTo(other.accountId);
    }
    if (result == 0) {
      result = containerId.compareTo(other.containerId);
    }
    if (result == 0) {
      result = partitionId.compareTo(other.partitionId);
    }
    if (result == 0) {
      result = uuid.compareTo(other.uuid);
    }
```

xiahome replied:
Ok, after a second thought, it seems the current code could avoid some minor overhead if `result==0` at early checks. It will not execute the rest of checks.

nsivabalan replied:
sorry i don't get you

xiahome replied:
Will explain to you.

nsivabalan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobIdBuilder.java:
```diff
@@ -0,0 +1,103 @@
+/**
+ * A builder class for {@link BlobId}. This class is not thread safe.
+ */
+public class BlobIdBuilder {
```
Had a discussion with @vgkholla . We feel that builder is not required here. Especially when we have two properly defined constructors and have mandatory fields. Can we remove the builder unless you have strong reasons to hold on to it.

xiahome replied:
Per our discussion offline, it is preferable to keep this builder for mainly two reasons:
1. We will have a single controller, that is the `CURRENT_VERSION`, to control the serialized version for `BlobId`s constructed by providing field arguments. We will not use constructors to control the version. So There won't be any code change if we want to change the serialization format;
2. The field can be null, and it is at the `BlobId` level to decide the values for missing fields. This will be better than the caller to know the default/legacy values for any N/A fields. Therefore, if the caller does not have the value for a missing field, it just supplies null or does not set the field, the `BlobId` will resolve it.

nsivabalan commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,538 @@
+    random.nextBytes(bytes);
+    referenceFlag = bytes[0];
+    referenceDatacenterId = bytes[0];
+    referenceAccountId = getRandomShort(random);
```
We already have getRandomShort() in Utils.

xiahome replied:
This is indeed a utils method.

nsivabalan commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,538 @@
+    }
+    BlobId blobId = blobIdBuilder.build();
+    assertEquals("Wrong blobId version", version, getVersionFromBlobString(blobId.getID()));
+    assertBlob(version, blobId, flag, datacenterId, accountId, containerId, partitionId, useSetter);
```
assertBlobIdFieldValues

xiahome replied:
Renamed.

nsivabalan commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,538 @@
+        buildBlobIdAndAssert(version, referenceFlag, referenceDatacenterId, referenceAccountId, referenceContainerId,
+            referencePartitionId, shouldUseSetter);
+        buildBlobIdAndAssert(version, null, referenceDatacenterId, referenceAccountId, referenceContainerId,
+            referencePartitionId, shouldUseSetter);
```
shouldn't we have non null values for all new fields in V2?

xiahome replied:
Replied in the 2nd point of the response to `Builder`.

nsivabalan commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,538 @@
+    switch (version) {
+      case BLOB_ID_V1:
+        srcBlobIdStr =
+            buildBlobIdLike(version, null, null, null, null, referencePartitionId, srcUUID.length(), srcUUID);
```
since we use instance variables for most part to construct BlobId, may be you can introduce a privateMethod which takes in non-instance variables and version to construct BlobId is the specified version

xiahome replied:
Actually this switch case is not needed, because the `buildBlobIdLike` method has the logic. For other switch occurrences, it seems most of them have specific inside-switch logics, and consolidating them may not save much. We can discuss over this if there is a different opinion.

nsivabalan commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,538 @@
+   * @param referencePartitionId The reference PartitionId.
+   * @throws Exception Any unexpected exception.
+   */
+  private void deserializeBlobIdAndAssert(short version, String srcBlobIdStr, PartitionId referencePartitionId)
```
referencePartitionId is a ref. why we need to pass explicitly

xiahome replied:
Actually you are right, this is not necessary and hence removed.

nsivabalan commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,538 @@
+   * @throws Exception Any unexpected exception.
+   */
+  private void generateAndAssertBadBlobId(Short version) throws Exception {
+    List<String> blobIdLikes = new ArrayList<>();
```
invalidBlobIds

xiahome replied:
Changed to `invalidBlobIdLikeList`. Prefer to keep `Like` because the ids are hand-made.

nsivabalan commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,538 @@
+        idBuf.putShort(containerId);
+        break;
+
+      default:
```
why not fail in un-supported versions ?

xiahome replied:
This will build some blobIdLike with a bad version. The bad-version id will fail a test later.

xiahome commented:
@pnarayanan @nsivabalan @vgkholla Have addressed your comments (for @vgkholla I have removed the `builder`), and rebased on using `BlobPropertiesV2`. Please take a look.

nsivabalan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,90 +27,253 @@
-    return (short) (Version_Size_In_Bytes + partitionId.getBytes().length + UUID_Size_In_Bytes + uuid.length());
+    switch (version) {
+      case BLOB_ID_V1:
+        return (short) (VERSION_FIELD_LENGTH_IN_BYTES + partitionId.getBytes().length + UUID_SIZE_FIELD_LENGTH_IN_BYTES
```
try to save the size in a variable. Incase of V1, you can return right away. Incase of V2, you can build upon the same.

xiahome replied:
Thought a bit and would rather keep the current way. Could not save many code nor increase readability.

xiahome replied:
Fixed.

nsivabalan commented on ambry-admin/src/test/java/com.github.ambry.admin/GetReplicasHandlerTest.java:
```diff
@@ -62,7 +65,8 @@ public void getReplicasTest() throws Exception {
     for (PartitionId partitionId : partitionIds) {
       String originalReplicaStr = partitionId.getReplicaIds().toString().replace(", ", ",");
-      BlobId blobId = new BlobId(partitionId);
+      BlobId blobId = new BlobId(BlobId.DEFAULT_FLAG, ClusterMapUtils.LEGACY_DATACENTER_ID, Account.LEGACY_ACCOUNT_ID,
```
I know I have a split opinion. I'd would take a stance that this constructor should be used only when we start writing in V2. Until then, its confusing why the callers are sending in defaults. This makes the caller assume that toBytes() will serialize in V2, which is not. We have to come up with some standardization wrt this new version upgrades in general.

xiahome replied:
I guess since our prod code already starts using this constructor, the tests should be the same. We have a notion that v1 and v2 are purely controlled by `BlobId#CURRENT_VERSION`, so no need to revisit every place (e.g., in this test) when there is a version change.

nsivabalan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -162,18 +354,34 @@ public boolean equals(Object o) {
+   * Gets the value of {@link #CURRENT_VERSION}.
+   * @return The value of {@link #CURRENT_VERSION}.
+   */
+  short getCurrentVersion() {
```
protected

xiahome replied:
added.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/PutOperation.java:
```diff
@@ -787,7 +788,9 @@ private void prepareForSending() {
         partitionId = getPartitionForPut(attemptedPartitionIds);
-        chunkBlobId = new BlobId(partitionId);
+        chunkBlobId =
+            new BlobId(blobIdFlag, clusterMap.getLocalDatacenterId(), passedInBlobProperties.getAccountId(),
```
why not use BlobId.DEFAULT_FLAG directly here.

xiahome replied:
ok, i will remove

nsivabalan commented on ambry-tools/src/main/java/com.github.ambry/tools/admin/DirectoryUploader.java:
```diff
@@ -128,7 +128,7 @@ private void setDataNodeId(ClusterMap clusterMap, String hostname, int port, boo
 
   public void walkDirectoryToCreateBlobs(String path, FileWriter writer, String datacenter,
-      boolean enableVerboseLogging) throws InterruptedException {
+      boolean enableVerboseLogging, ClusterMap map) throws InterruptedException {
```
you could directly send datacenterId here instead of clustermap

xiahome replied:
done.

nsivabalan commented on ambry-tools/src/main/java/com.github.ambry/tools/perf/ServerWritePerformance.java:
```diff
@@ -336,12 +336,15 @@ public void run() {
           byte[] usermetadata = new byte[new Random().nextInt(1024)];
           BlobProperties props = new BlobProperties(randomNum, "test");
           ConnectedChannel channel = null;
+          byte blobFlag = BlobId.DEFAULT_FLAG;
```
we can remove the declaration here and use it directly

xiahome replied:
Removed.

nsivabalan commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,447 @@
-    PartitionId partitionId = new MockPartitionId(id, Collections.EMPTY_LIST, 0);
-    BlobId blobId = new BlobId(partitionId);
+  public void testBuildBlobId() throws Exception {
+    for (Short version : versions) {
```
if most of you test cases runs for both version, why not go our usual route of parametrized args.

xiahome replied:
Changed to parameterized args.

nsivabalan commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,447 @@
-    final int idLength = 2 + partitionId.getBytes().length + 4 + uuidLike.length();
-    ByteBuffer idBuf = ByteBuffer.allocate(idLength);
-    idBuf.putShort(version);
+  private String buildBlobIdLike(short version, Byte flag, Byte datacenterId, Short accountId, Short containerId,
```
why we need this method. Why can't we use actual BlobId class? The way we solved for one my previous patch is, we had a protected constructor which takes in a version. Rest of the code should be as is

xiahome replied:
This is to mimic some fake ids, because blobid strings are not necessarily generated from `BlobId`, but can be maliciously made in various format. This is to ensure out deserialization is working well under malicious inputs.

nsivabalan commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,447 @@
+      assertEquals("Wrong base-64 ID in blobId: " + blobId, srcBlobIdStr, blobId.getID());
+      assertEquals("Wrong blobId version", version, getVersionFromBlobString(blobId.getID()));
+      switch (version) {
+        case BLOB_ID_V1:
```
is it possible to re-use assertBlobIdFieldValues() here

xiahome replied:
Reused.

nsivabalan commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,447 @@
+   * @throws Exception Any unexpected exception.
+   */
+  @Test
+  public void testDeserialization() throws Exception {
```
I don't understand the need of this test. We have the next one which tests serialization and deserialization if I am not wrong. Or does this tests something more than the next one ?

xiahome replied:
Please see my response to your next comment.

vgkholla commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,90 +27,253 @@
+  private static final short ACCOUNT_ID_FIELD_LENGTH_IN_BYTES = Short.BYTES;
+  private static final short CONTAINER_ID_FIELD_LENGTH_IN_BYTES = Short.BYTES;
+  // the version to indicate the serialized format.
+  private Short version;
```
these could all be `final`?

xiahome replied:
Added.

vgkholla commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,90 +27,253 @@
+      throw new IllegalArgumentException("partitionId cannot be null");
+    }
+    version = getCurrentVersion();
+    switch (version) {
```
not sure why this switch case is required - just use whatever is passed in?

xiahome replied:
It depends on how to treat the passed-in values if they are not required. For example, if I build a `blobId` with `flag=3`, but `CURRENT_VERSION` is 1, then what should we expect when we call `blobId.getFlag()`. We can either return 3, or `DEFAULT_FLAG`. No one is absolutely correct or wrong, but I am inclined to return the latter, because this will provide consistent result after serdes.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobIdFactory.java:
```diff
@@ -33,6 +33,6 @@ public BlobIdFactory(ClusterMap clusterMap) {
   @Override
   public StoreKey getStoreKey(DataInputStream value) throws IOException {
-    return new BlobId(value, clusterMap);
+    return BlobId.fromDataInputStream(value, clusterMap);
```
There was some discussion that there should be a `static` method that does the des job, and building from scratch using constructor. I guess no such rule exists, so I will change it to constructor.

xiahome commented on ambry-admin/src/test/java/com.github.ambry.admin/AdminBlobStorageServiceTest.java:
```diff
@@ -389,7 +392,8 @@ public void getReplicasTest() throws Exception {
     for (PartitionId partitionId : partitionIds) {
       String originalReplicaStr = partitionId.getReplicaIds().toString().replace(", ", ",");
-      BlobId blobId = new BlobId(partitionId);
+      BlobId blobId = new BlobId(BlobId.DEFAULT_FLAG, ClusterMapUtils.LEGACY_DATACENTER_ID, Account.LEGACY_ACCOUNT_ID,
```
Good catch, thanks! As discussed with Siva before, they were temporarily set in `BlobProperties`. Will remove them.

xiahome replied:
Removed.

xiahome commented on ambry-admin/src/main/java/com.github.ambry.admin/GetReplicasHandler.java:
```diff
@@ -86,7 +86,7 @@ ReadableStreamChannel getReplicas(String blobId, RestResponseChannel restRespons
   private JSONObject getReplicas(String blobId) throws RestServiceException {
     try {
-      PartitionId partitionId = new BlobId(blobId, clusterMap).getPartition();
+      PartitionId partitionId = BlobId.fromStringId(blobId, clusterMap).getPartition();
```
There was some discussion that there should be a static method that does the des job, and building from scratch using constructor. I guess no such rule exists, so I will change it to constructor.

xiahome commented:
@pnarayanan @nsivabalan @vgkholla Have updated the pr.

vgkholla commented on ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java:
```diff
@@ -30,7 +30,11 @@
 import com.github.ambry.messageformat.BlobAll;
 import com.github.ambry.messageformat.BlobProperties;
 import com.github.ambry.messageformat.MessageFormatException;
+import com.github.ambry.network.ChannelOutput;
```
why are all these imports added even though there is no code added? Has everything been formatted?

xiahome replied:
Removed and a global format is made.

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,440 @@
+   */
+  @Before
+  public void init() throws Exception {
+    byte[] bytes = new byte[2];
```
why not set all of these in the constructor? Maybe you can then make most class variables `final`

xiahome replied:
Moved all to constructor.

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,440 @@
+        assertEquals("Wrong container id in blobId: " + blobId, containerId, blobId.getContainerId());
+        break;
+    }
+    System.out.println("BlobId=" + blobId + ", flag=" + blobId.getFlag() + ", datacenterId=" + blobId.getDatacenterId()
```
remove System.out

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,440 @@
+ * in v2 equivalent to setting {@link BlobId#CURRENT_VERSION} to be {@link BlobId#BLOB_ID_V2}.
+ */
+@RunWith(Parameterized.class)
 public class BlobIdTest {
```
perhaps I missed it but can we have a test that ser-deser-ser and make sure the first and last ser match?

xiahome replied:
This is truly a useful test, and it is in Line 291
`assertEquals("Wrong base-64 ID in blobId: " + blobId, srcBlobIdStr, blobId.getID());`, where the `srcBlobIdStr` is a serialized string, and `blobId.getID()` is a string after deser-ser.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,35 +27,153 @@
    */
-  public BlobId(PartitionId partitionId) {
+  private BlobId(DataInputStream stream, ClusterMap clusterMap, boolean ensureFullyRead) throws IOException {
+    this.version = stream.readShort();
```
Removed.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -62,52 +183,97 @@ public BlobId(String id, ClusterMap clusterMap) throws IOException {
-    return (short) (Version_Size_In_Bytes + partitionId.getBytes().length + UUID_Size_In_Bytes + uuid.length());
+    short sizeForBlobIdV1 =
+        (short) (VERSION_FIELD_LENGTH_IN_BYTES + partitionId.getBytes().length + UUID_SIZE_FIELD_LENGTH_IN_BYTES
+            + uuid.length());
```
`length()` has been used since start. If `uuid.getBytes().length` produces the same result, it is ok we change, otherwise we have to be careful.

xiahome replied:
Got your point, will fix now.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -62,52 +183,97 @@ public BlobId(String id, ClusterMap clusterMap) throws IOException {
 
+  /**
+   * Gets the id of the {@link Account} who created this blob. If this information was not available when the
+   * blobId was formed, it will return -1.
```
Fixed this and `getContainerId` and `getDatacenterId` and `getFlag`.

xiahome replied:
Also one change to be made is changing `LEGACY_XXX_ID` to `UNKNOWN_XXX_ID`. This is because these ids mean no known id information could be obtained when forming the corresponding fields.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -62,52 +183,97 @@ public BlobId(String id, ClusterMap clusterMap) throws IOException {
+
+  /**
+   * Gets the flag metadata of this blobId. If this information was not available when the blobId was formed, it
+   * will return 0.
```
fixed.

xiahome commented:
@pnarayanan @nsivabalan @vgkholla Have updated the PR.

pnarayanan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,35 +27,153 @@
+        accountId = UNKNOWN_ACCOUNT_ID;
+        containerId = UNKNOWN_CONTAINER_ID;
+        break;
+
```
extra line.

xiahome replied:
Removed all the extra lines.

pnarayanan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,35 +27,153 @@
+        accountId = stream.readShort();
+        containerId = stream.readShort();
+        break;
+
```
extra line.

pnarayanan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -62,52 +183,97 @@ public BlobId(String id, ClusterMap clusterMap) throws IOException {
+   * when the blobId was formed, it will return {@link ClusterMapUtils#UNKNOWN_DATACENTER_ID}.
+   * @return The id of the datacenter where this blob was originally posted.
+   */
+  public short getDatacenterId() {
```
why isn't the return value a `byte`?

xiahome replied:
Changed to `byte`

pnarayanan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -62,52 +183,97 @@ public BlobId(String id, ClusterMap clusterMap) throws IOException {
+   * will return {@link #DEFAULT_FLAG}.
+   * @return The flag of the blobId.
+   */
+  public short getFlag() {
```
why isn't the return value a `byte`?

xiahome replied:
changed to `byte`

pnarayanan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -140,9 +320,21 @@ public int compareTo(StoreKey o) {
 
     int result = version.compareTo(other.version);
     if (result == 0) {
```
These extra information that we are adding (account id, flags etc.) do not necessarily have to get accounted in the comparison. This method could actually go untouched as you will not have blob ids that just differ in their flags or account ids.

This is probably okay - but do make sure that the comparison stays exactly the same for V1 blob ids (it seems to), otherwise the lookups in existing segments will go wrong.

xiahome replied:
Sure, i just feel a bit hesitate to treat two blobs the same if all fields are the same but flag different. Would rather prefer this more rigorous way but of cause need to ensure that this is compatible for v1 blobs.

xiahome replied:
Have changed to use a `switch`, for V1 we just use the old method (keep the old code), and for V2 we compare all the fields. This will ensure that comparison among all existing blobids will not change.

pnarayanan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -162,18 +354,34 @@ public boolean equals(Object o) {
     if (!version.equals(blobId.version)) {
```
`compareTo()` and `equals()` have to be consistent with each other, which it is here. But you can enforce it and simplify by simply doing `return compareTo(blobId) == 0;` at line 353 and avoiding the code duplication.

xiahome replied:
I see. I like your idea.

pnarayanan commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,429 @@
+      case BLOB_ID_V1:
+        blobId = new BlobId(flag, datacenterId, accountId, containerId, partitionId);
+        break;
+
```
extra line

pnarayanan commented on ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java:
```diff
@@ -13,119 +13,429 @@
+      return BLOB_ID_V2;
+    }
+  }
 }
```
since you modified the storeKey `compareTo()` and `equals()` methods and since we need to make sure that the behavior of these methods remains the same between the new code and the old code (which was used to sort and write these ids in the on-disk index segments), is it possible to add a few test cases to verify that?

xiahome replied:
Will add.

xiahome replied:
Added two tests.

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -62,52 +183,97 @@ public BlobId(String id, ClusterMap clusterMap) throws IOException {
+    switch (version) {
+      case BLOB_ID_V1:
+        return sizeForBlobIdV1;
+
```
Removed all extra lines in `switch`

xiahome commented:
@pnarayanan Have addressed your comments.

xiahome commented:
@pnarayanan Removed the test.

pnarayanan commented on ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java:
```diff
@@ -24,35 +27,149 @@
+    if (partitionId == null) {
+      throw new IllegalArgumentException("partitionId cannot be null");
+    }
+    version = getCurrentVersion();
```
I know I had suggested overriding `getCurrentVersion()` to test V2, but I did not mean for this to be called from the constructor. Constructors invoking overridable methods is generally considered bad form. `BlobIdV2` in the tests that extends `BlobId` is not completely constructed when the superclass constructor here calls into `BlobIdV2::getCurrentVersion()`. If this method was accessing object state then it would lead to undefined behavior (was okay here because it just returns a `static final` variable)

This can be avoided by doing this check within the other methods (that is instead of checking `if (version == x)`, those other methods would do `if (getCurrentVersion() == x)`.

xiahome replied:
Will put up a PR to address this issue. Thanks for pointing out.

xiahome replied:
As discussed offline, the suggestion is generally correct, but could be a bit tricky to apply to this case. The reason is there is a `version` field that needs to be set during BlobId construction, and it may be set by either `CURRENT_VERSION` or from input stream. This makes it difficult to call `getCurrentVersion` every place when `version` is needed. After discussion, we decided to keep the current way, but will add javadoc to note that this is an unusual paradigm.
